File: dart/lib/authdog_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import 'exceptions.dart';
import 'types.dart';

class AuthdogClientConfig {
  final String baseUrl;
  final String? apiKey;
  final Duration timeout;
  final http.Client? httpClient;

  const AuthdogClientConfig({
    required this.baseUrl,
    this.apiKey,
    this.timeout = const Duration(seconds: 10),
    this.httpClient,
  });
}

class AuthdogClient {
  final String _baseUrl;
  final String? _apiKey;
  final http.Client _client;
  final Duration _timeout;

  AuthdogClient(AuthdogClientConfig config)
      : _baseUrl = config.baseUrl.replaceAll(RegExp(r"/+$$"), ''),
        _apiKey = config.apiKey,
        _timeout = config.timeout,
        _client = config.httpClient ?? http.Client();

  Map<String, String> _defaultHeaders() {
    final headers = <String, String>{
      'Content-Type': 'application/json',
      'User-Agent': 'authdog-dart-sdk/0.1.0',
    };
    if (_apiKey != null && _apiKey!.isNotEmpty) {
      headers['Authorization'] = 'Bearer ' + _apiKey!;
    }
    return headers;
  }

  Future<UserInfoResponse> getUserInfo(String accessToken) async {
    final uri = Uri.parse(_baseUrl + '/v1/userinfo');
    final headers = Map<String, String>.from(_defaultHeaders());
    headers['Authorization'] = 'Bearer ' + accessToken;

    try {
      final response = await _client
          .get(uri, headers: headers)
          .timeout(_timeout);

      final status = response.statusCode;
      final body = response.body;

      if (status == 401) {
        throw AuthenticationException('Unauthorized - invalid or expired token');
      }

      if (status == 500) {
        try {
          final error = json.decode(body) as Map<String, dynamic>;
          final err = error['error'];
          if (err == 'GraphQL query failed') {
            throw ApiException('GraphQL query failed');
          } else if (err == 'Failed to fetch user info') {
            throw ApiException('Failed to fetch user info');
          }
        } catch (_) {}
      }

      if (status < 200 || status >= 300) {
        throw ApiException('HTTP error ' + status.toString() + ': ' + body);
      }

      final data = json.decode(body) as Map<String, dynamic>;
      return UserInfoResponse.fromJson(data);
    } on AuthdogException {
      rethrow;
    } on http.ClientException catch (e) {
      throw ApiException('Request failed: ' + e.toString());
    } on FormatException catch (e) {
      throw ApiException('Failed to parse response: ' + e.message);
    } catch (e) {
      throw ApiException('Request failed: ' + e.toString());
    }
  }

  void close() {
    _client.close();
  }
}

File: fsharp/src/AuthdogClient.fs
namespace Authdog

open System
open System.Net.Http
open System.Text.Json
open System.Threading.Tasks

[<CLIMutable>]
type AuthdogClientConfig =
    { BaseUrl: string
      ApiKey: string option
      TimeoutMs: int option
      HttpClient: HttpClient option }

module private Internal =
    let trimTrailingSlashes (s: string) = s.TrimEnd([|'/'|])

    let addDefaultHeaders (apiKey: string option) (req: HttpRequestMessage) =
        req.Headers.TryAddWithoutValidation("Content-Type", "application/json") |> ignore
        req.Headers.TryAddWithoutValidation("User-Agent", "authdog-fsharp-sdk/0.1.0") |> ignore
        match apiKey with
        | Some key when not (String.IsNullOrWhiteSpace key) ->
            req.Headers.Remove("Authorization") |> ignore
            req.Headers.TryAddWithoutValidation("Authorization", "Bearer " + key) |> ignore
        | _ -> ()

/// Minimal Authdog API client
[<Sealed>]
type AuthdogClient (config: AuthdogClientConfig) =
    let baseUrl = Internal.trimTrailingSlashes config.BaseUrl
    let apiKey = config.ApiKey
    let httpClient =
        match config.HttpClient with
        | Some c -> c
        | None ->
            let c = new HttpClient()
            c.Timeout <- TimeSpan.FromMilliseconds(float (defaultArg config.TimeoutMs 10000))
            c

    member _.GetUserInfo(accessToken: string) : Task<string> = task {
        let url = baseUrl + "/v1/userinfo"
        use req = new HttpRequestMessage(HttpMethod.Get, url)
        Internal.addDefaultHeaders apiKey req
        let authToUse =
            match apiKey with
            | Some key when not (String.IsNullOrWhiteSpace key) -> key
            | _ -> accessToken
        req.Headers.Remove("Authorization") |> ignore
        req.Headers.TryAddWithoutValidation("Authorization", "Bearer " + authToUse) |> ignore
        try
            use! resp = httpClient.SendAsync(req)
            let! body = resp.Content.ReadAsStringAsync()
            if resp.StatusCode = System.Net.HttpStatusCode.Unauthorized then
                return raise (AuthenticationException("Unauthorized - invalid or expired token"))
            elif int resp.StatusCode = 500 then
                try
                    use doc = JsonDocument.Parse(body)
                    let root = doc.RootElement
                    if root.TryGetProperty("error", ref (Unchecked.defaultof<JsonElement>)) then
                        let err = root.GetProperty("error").GetString()
                        match err with
                        | "GraphQL query failed" -> return raise (ApiException("GraphQL query failed"))
                        | "Failed to fetch user info" -> return raise (ApiException("Failed to fetch user info"))
                        | _ -> ()
                with _ -> ()
            if not resp.IsSuccessStatusCode then
                return raise (ApiException(sprintf "HTTP error %d: %s" (int resp.StatusCode) body))
            return body
        with
        | :? AuthenticationException as e -> return raise e
        | :? ApiException as e -> return raise e
        | e -> return raise (ApiException("Request failed: " + e.Message))
    }
